File: nodejs/tests/public_api_live.test.js
// Copyright 2026, Pulumi Corporation.  All rights reserved.

// Live end-to-end tests against unauthenticated Pulumi Cloud routes. Mirrors
// sdk/python/tests/test_public_api_live.py. Real HTTP status errors fail;
// connectivity failures (offline) skip so the suite stays green without network.
// Override the host with PULUMI_API_HOST (base URL, e.g. https://api.staging...).

const { test } = require("node:test");
const assert = require("node:assert/strict");

const { ApiService, ApiClient, ApiClientConfiguration, ApiError } = require("../dist/index.js");

function newApis() {
    const base = process.env.PULUMI_API_HOST;
    // apiHost must include the /api/ prefix and a trailing slash.
    const apiHost = base ? `${base.replace(/\/+$/, "")}/api/` : "https://api.pulumi.com/api/";
    const config = new ApiClientConfiguration(apiHost, "pulumi-cloud-sdk-test");
    return new ApiService(new ApiClient(config));
}

// A connectivity failure surfaces as a plain fetch TypeError (not an ApiError);
// treat that as "offline" and skip. A non-2xx HTTP status is a real failure.
function isConnectivityError(err) {
    return !(err instanceof ApiError);
}

test("GET /api/cli/version returns a semver-shaped latestVersion", async (t) => {
    let version;
    try {
        version = await newApis().Miscellaneous.Version();
    } catch (err) {
        if (isConnectivityError(err)) {
            t.skip(`Pulumi Cloud unreachable: ${err}`);
            return;
        }
        throw err;
    }
    assert.match(String(version.latestVersion), /^\d+\.\d+\.\d+/);
});

test("GET /api/capabilities returns a capabilities list", async (t) => {
    let capabilities;
    try {
        capabilities = await newApis().Miscellaneous.Capabilities();
    } catch (err) {
        if (isConnectivityError(err)) {
            t.skip(`Pulumi Cloud unreachable: ${err}`);
            return;
        }
        throw err;
    }
    assert.ok(Array.isArray(capabilities.capabilities));
});

File: go/tests/networktests/public_api_live_test.go
// Copyright 2026, Pulumi Corporation.  All rights reserved.

// A live, unauthenticated call against the public Pulumi Cloud API. Ports
// sdk/java/src/test/java/com/pulumi/cloud/sdk/PublicApiLiveTest.java and
// sdk/dotnet/Pulumi.Cloud.Sdk.Tests/PublicApiLiveTest.cs, and mirrors
// sdk/python/tests/test_public_api_live.py and
// sdk/nodejs/tests/public_api_live.test.js.
//
// This is a smoke test of the whole stack — real DNS, TLS, routing, and JSON
// decoding of a real response — which no amount of stubbed-Executor testing
// covers.
//
// It lives in its own package, and its own Bazel target, so that it can carry
// `requires-network` without dragging the offline suite in sdk/go/tests onto a
// network-enabled worker.
//
// Like its four counterparts it *skips* rather than fails when the request never
// reaches the server: a developer offline, or an RBE worker without egress,
// should not see a red test. A real HTTP status is a genuine failure and is
// reported as one.
package networktests

import (
	"context"
	"errors"
	"net/http"
	"os"
	"regexp"
	"testing"

	"github.com/pulumi/pulumi-cloud-sdk/go/apiclient"
)

var semver = regexp.MustCompile(`^\d+\.\d+\.\d+`)

func TestVersionReturnsSemver(t *testing.T) {
	t.Parallel()

	host := os.Getenv("PULUMI_API_HOST")
	if host == "" {
		host = "https://api.pulumi.com"
	}

	client := &apiclient.CloudClient{
		BaseURL:  host,
		Executor: http.DefaultClient.Do,
	}

	response, err := client.Version(context.Background())
	if err != nil {
		// An APIError means we reached the service and it answered — that is a
		// real failure. Anything else is a transport problem; skip.
		var apiErr *apiclient.APIError
		if errors.As(err, &apiErr) {
			t.Fatalf("Version returned HTTP %d: %s", apiErr.HTTPStatusCode(), apiErr.ResponseMessage())
		}
		t.Skipf("skipping: could not reach %s: %v", host, err)
	}

	if response == nil {
		t.Fatal("Version returned a nil response and no error")
	}
	if !semver.MatchString(response.LatestVersion) {
		t.Errorf("LatestVersion: got %q, want a semver string", response.LatestVersion)
	}
}
